dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/task/AbstractJdbcTask.java

```diff
@@ -76,28 +84,68 @@ public AbstractJdbcTask<T> withHeaderClass(@Nonnull Class<? extends Enum<?>> hea
     return this;
   }
 
+  @Nonnull
+  public AbstractJdbcTask<T> withAssessment(boolean flag) {
+    this.isAssessment = flag;
+    return this;
+  }
+
+  @Nonnull
+  public AbstractJdbcTask<T> withHeaderValidation(boolean isRequired) {
+    this.requiresHeaderValidation = isRequired;
+    return this;
+  }
+
   @Nonnull
   protected CSVFormat newCsvFormat(@Nonnull ResultSet rs) throws SQLException {
     CSVFormat format = FORMAT;
     Class<? extends Enum<?>> headerClass = getHeaderClass();
     if (headerClass != null) {
       format = format.withHeader(headerClass);
-      if (headerClass.getEnumConstants().length != rs.getMetaData().getColumnCount())
-        // Can we avoid nesting exceptions here?
-        throw new SQLException(
-            new MetadataDumperUsageException(
-                "Fatal Error. ResultSet does not have the expected column count: "
-                    + headerClass.getEnumConstants().length,
-                Arrays.asList(
-                    "If a custom query has been specified please confirm the selected columns match"
-                        + " the following: ",
-                    StringUtils.join(headerClass.getEnumConstants(), ", "))));
+      ResultSetMetaData metaData = rs.getMetaData();
+      validateHeader(headerClass, metaData);
     } else {
       format = format.withHeader(rs);
     }
     return format;
   }
 
+  private void validateHeader(Class<? extends Enum<?>> headerClass, ResultSetMetaData metaData)
+      throws SQLException {
+    if (isAssessment && requiresHeaderValidation) {
+
+      Set<String> resultColumnNames = new HashSet<>();
+
+      for (int i = 0; i < metaData.getColumnCount(); i++) {
+        resultColumnNames.add(metaData.getColumnName(i + 1));
+      }
+
+      Set<String> desiredColumnNames =
+          Arrays.stream(headerClass.getEnumConstants())
+              .map(
+                  enumVal -> CaseFormat.UPPER_CAMEL.to(CaseFormat.UPPER_UNDERSCORE, enumVal.name()))
+              .collect(Collectors.toSet());
+
+      desiredColumnNames.removeAll(resultColumnNames);
+
+      if (!desiredColumnNames.isEmpty()) {
+        throw new SQLException(
+            new MetadataDumperUsageException(
+                "Fatal Error. ResultSet does not have the expected columns: "
+                    + desiredColumnNames));
+      }
+    } else if (headerClass.getEnumConstants().length != metaData.getColumnCount())
+      // Can we avoid nesting exceptions here?
+      throw new SQLException(
```

Comment:
Aren't we supposed to not throw exception in this case?

Reply:
oops! my bad. should be okay now